numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -328,28 +328,31 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
+    PyObject  *item, *item_str;
 
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    item = gentype_generic_method(self, NULL, NULL, "item");
+    if (item == NULL) {
+        return NULL;
     }
-    return ret;
+
+    item_str = PyObject_Str(item);
```

Comment:
Right, you are talking about [this](https://github.com/numpy/numpy/pull/9143#discussion_r118132580) comment.

Currently, besides potential user-defined types, only numpy only void types fall back to this. But they no longer will after #8981.

I just tried you suggestion by simply removing `gentype_repr` and `gentype_str`, in the last commit pushed here.

Locally (not in the pushed commit) I also temporarily modified `test_rational.c` to remove the str and repr of rational types as a check, and can confirm I get reasonable behavior:
```python
>>> from numpy.core.test_rational import rational
>>> a = np.array([1,2,3], dtype=rational)
>>> a
array([<rational object at 0x7f37f1b4e918>,
       <rational object at 0x7f37f1b4e918>,
       <rational object at 0x7f37f1b4e918>], dtype=rational)
```

I'm not sure how to implement a test of this without recreating a new `test_rational` type missing the str and repr, which seems a bit heavy-handed.